add doc for users to migrate from oam-runtime to kubevela (#697)
oam-dev/kubevela: design/vela-core/migrate-from-oam-runtime.md

> # Migrate from OAM Runtime

Reviewer: ```suggestion
# Migrate from OAM Kubernetes Runtime
```

> # Migrate from OAM Kubernetes Runtime

Reviewer 2: # How to upgrade to KubeVela

> Now we are refactoring OAM runtime in vela-core with CUE based abstractions, all source code from [oam-kubernetes-runtime](https://github.com/crossplane/oam-kubernetes-runtime)

Reviewer: ```suggestion
Now we are refactoring OAM runtime in vela-core with CUE based abstractions. All source code from [oam-kubernetes-runtime](https://github.com/crossplane/oam-kubernetes-runtime)
```

> Now we are refactoring OAM runtime in vela-core with CUE based abstractions. All source code from [oam-kubernetes-runtime](https://github.com/crossplane/oam-kubernetes-runtime)

Reviewer 2: ```suggestion
What if I want to upgrade from [oam-kubernetes-runtime](https://github.com/crossplane/oam-kubernetes-runtime) to KubeVela? Here's a detailed guide!
```

> has already merged into [vela-core](https://github.com/oam-dev/kubevela/pull/663) now. Here is the doc for users who want

Reviewer: ```suggestion
has already been merged into [vela-core](https://github.com/oam-dev/kubevela/pull/663) now. Here is the doc for users who want
```

to migrate from OAM Runtime to use KubeVela.

## For Users Who are Using OAM runtime as Standalone Controller

If you are using OAM Runtime as Standalone Controller, migrating to KubeVela is very straight forward.

> Server-side KubeVela(We call it `vela-core` for convenience) now include following **BUILT-IN** CRD and controllers.

Reviewer: ```suggestion
Server-side KubeVela(We call it `vela-core` for convenience) now includes following **BUILT-IN** CRDs and controllers.
```

| Type |  CRD   | Controller  | From |
| ---- |  ----  | ----  | ----  |

> | Framework | `applicationconfigurations.core.oam.dev` | Yes | OAM Runtime |

Reviewer: Why do we call it `Framework`?

Author: Do you have any suggestions?

Reviewer: Nope, I'm just curious about the meaning.

Reviewer 2: Control Plane Objects

| Framework | `components.core.oam.dev` | No | OAM Runtime |

> | Workload | `containerizedworklaods.core.oam.dev` | Yes | OAM Runtime |

Reviewer 2: Workload Type

| Scope | `healthscope.core.oam.dev` | Yes | OAM Runtime |
| Trait | `manualscalertraits.core.oam.dev` | Yes | OAM Runtime |

> | Framework | `scopedefinitions.core.oam.dev` | No | OAM Runtime |

Reviewer 2: Control Plane Object, ditto to others

Author: fixed

| Framework | `traitdefinitions.core.oam.dev` | No | OAM Runtime |
| Framework | `workloaddefinitions.core.oam.dev` | No | OAM Runtime |
| Trait | `autoscalers.standard.oam.dev` | Yes | New in KubeVela |
| Trait | `metricstraits.standard.oam.dev` | Yes | New in KubeVela |
| Workload | `podspecworkloads.standard.oam.dev` | Yes | New in KubeVela |
| Trait | `route.standard.oam.dev` | Yes | New in KubeVela |

> CRDs and Controllers in the table from 'OAM Runtime' are exactly the same ones in `oam-kubernetes-runtime`.

Reviewer: ```suggestion
CRDs and Controllers in the table from 'OAM Runtime' are exactly the same to ones in `oam-kubernetes-runtime`.
```

So in KubeVela we have added 4 more new CRDs with controller. 

### Option 1: You want only Pure OAM Runtime by using vela-core with No additional traits and workloads.

1. Find you deployment

```shell script
$ kubectl -n oam-system get deployment -l app.kubernetes.io/name=oam-kubernetes-runtime
NAME                         READY   UP-TO-DATE   AVAILABLE   AGE
oam-kubernetes-runtime-oam   1/1     1            1           62s
```

2. Update the deployment

In this case, the deployment name of OAM runtime is `oam-kubernetes-runtime-oam`, let's edit it to update the image:

```shell script
$ kubectl -n oam-system edit deployment oam-kubernetes-runtime-oam
```

There are two changes:

- update the image from `crossplane/oam-kubernetes-runtime:latest` to `oamdev/vela-core:latest`

> - add an args `- "--disable-caps=all"`, this will disable all additional workload and trait built in vela-core.

Reviewer: ```suggestion
- add an args `- "--disable-caps=all"`, which will disable all additional workloads and traits built in vela-core.
```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: oam-kubernetes-runtime-oam
...
      containers:
        - name: oam
          args:
            - "--metrics-addr=:8080"
            - "--enable-leader-election"
+           - "--disable-caps=all"
-         image: crossplane/oam-kubernetes-runtime:latest
+         image: oamdev/vela-core:latest
          imagePullPolicy: "Always"
...
```

> ### Option 2: You want use fully functioned KubeVela

Reviewer: Why don't we directly delete the old release and install KubeVela chart intead?

Author: That's the easiest way, and it couldn't not be called by "migrate"

| Type |  CRD   | Controller  | From |
| ---- |  ----  | ----  | ----  |
| Control Plane Object | `applicationconfigurations.core.oam.dev` | Yes | OAM Runtime |
| Control Plane Object | `components.core.oam.dev` | Yes | OAM Runtime |
| Workload Type | `containerizedworklaods.core.oam.dev` | Yes | OAM Runtime |
| Scope | `healthscope.core.oam.dev` | Yes | OAM Runtime |
| Trait | `manualscalertraits.core.oam.dev` | Yes | OAM Runtime |
| Control Plane Object | `scopedefinitions.core.oam.dev` | No | OAM Runtime |
| Control Plane Object | `traitdefinitions.core.oam.dev` | No | OAM Runtime |
| Control Plane Object | `workloaddefinitions.core.oam.dev` | No | OAM Runtime |
| Trait | `autoscalers.standard.oam.dev` | Yes | New in KubeVela |
| Trait | `metricstraits.standard.oam.dev` | Yes | New in KubeVela |
| Workload Type | `podspecworkloads.standard.oam.dev` | Yes | New in KubeVela |
| Trait | `route.standard.oam.dev` | Yes | New in KubeVela |

CRDs and Controllers in the table from 'OAM Runtime' are exactly the same to those in `oam-kubernetes-runtime`.
So in KubeVela we have added 4 more new CRDs with controller. 

> ### Option 1: I only want to have OAM control plane objects onlywithout additional traits and workload types.

Reviewer 2: ```suggestion
### Option 1: I only want to have OAM control plane objects only, no additional traits and workload types.
```

```

2. Update the deployment

In this case, the deployment name of OAM runtime is `oam-kubernetes-runtime-oam`, let's edit it to update the image:

```shell script
$ kubectl -n oam-system edit deployment oam-kubernetes-runtime-oam
```

There are two changes:

- update the image from `crossplane/oam-kubernetes-runtime:latest` to `oamdev/vela-core:latest`
- add an args `- "--disable-caps=all"`, which will disable all additional workloads and traits built in vela-core described in the following table.

| Type | Current KubeVela Additional CRD   |
| ---- |  ----  |
| Trait | `autoscalers.standard.oam.dev` |
| Trait | `metricstraits.standard.oam.dev` |
| Workload | `podspecworkloads.standard.oam.dev` |
| Trait | `route.standard.oam.dev` |

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: oam-kubernetes-runtime-oam
...
      containers:
        - name: oam
          args:
            - "--metrics-addr=:8080"
            - "--enable-leader-election"
+           - "--disable-caps=all"
-         image: crossplane/oam-kubernetes-runtime:latest
+         image: oamdev/vela-core:latest
          imagePullPolicy: "Always"
...
```

> ### Option 2: I want full functioning KubeVela, including its built-in workload types and traits.

Reviewer 3: ```suggestion
### Option 2: I want full featured KubeVela, including its built-in workload types and traits.
```

# How to upgrade to KubeVela

What if I want to upgrade from [oam-kubernetes-runtime](https://github.com/crossplane/oam-kubernetes-runtime) to KubeVela? Here's a detailed guide!

> has already been merged into [vela-core](https://github.com/oam-dev/kubevela/pull/663) now. Here is the doc for users who want

Reviewer 2: Delete the unneeded lines.

> has already been merged into [vela-core](https://github.com/oam-dev/kubevela/pull/663) now. Here is the doc for users who want

Reviewer 3: What has been merged?

Author: this line will be deleted

to migrate from OAM Runtime to use KubeVela.

> ## For Users Who are Using OAM runtime as Standalone Controller

Reviewer 3: We need to either capitalize all except prep. or only keep the first capitalized.
I would recommend the latter.

Author: sure

> If you are using OAM Runtime as Standalone Controller, migrating to KubeVela is very straight forward.

Reviewer 2: ```suggestion
If you are using OAM Runtime as Standalone Controller, upgrading to KubeVela to KubeVela is very straight forward.
```

| Type |  CRD   | Controller  | From |
| ---- |  ----  | ----  | ----  |
| Control Plane Object | `applicationconfigurations.core.oam.dev` | Yes | OAM Runtime |
| Control Plane Object | `components.core.oam.dev` | Yes | OAM Runtime |
| Workload | `containerizedworklaods.core.oam.dev` | Yes | OAM Runtime |
| Scope | `healthscope.core.oam.dev` | Yes | OAM Runtime |
| Trait | `manualscalertraits.core.oam.dev` | Yes | OAM Runtime |
| Control Plane Object | `scopedefinitions.core.oam.dev` | No | OAM Runtime |
| Control Plane Object | `traitdefinitions.core.oam.dev` | No | OAM Runtime |
| Control Plane Object | `workloaddefinitions.core.oam.dev` | No | OAM Runtime |
| Trait | `autoscalers.standard.oam.dev` | Yes | New in KubeVela |
| Trait | `metricstraits.standard.oam.dev` | Yes | New in KubeVela |
| Workload | `podspecworkloads.standard.oam.dev` | Yes | New in KubeVela |
| Trait | `route.standard.oam.dev` | Yes | New in KubeVela |

CRDs and Controllers in the table from 'OAM Runtime' are exactly the same to those in `oam-kubernetes-runtime`.
So in KubeVela we have added 4 more new CRDs with controller. 

> ### Option 1: You want only Pure OAM Runtime by using vela-core with No additional traits and workloads.

Reviewer 2: ```suggestion
### Option 1: I only want to have OAM control plane objects only,  no KubeVela traits and workloads.
```

> ### Option 1: You want only Pure OAM Runtime by using vela-core with No additional traits and workloads.

Reviewer 3: ```suggestion
### Option 1: You want only pure OAM Runtime without additional traits and workloads.
```

1. Find you deployment

```shell script
$ kubectl -n oam-system get deployment -l app.kubernetes.io/name=oam-kubernetes-runtime
NAME                         READY   UP-TO-DATE   AVAILABLE   AGE
oam-kubernetes-runtime-oam   1/1     1            1           62s
```

> 2. Update the deployment

Reviewer 3: Why not just delete OAM Runtime deployment and install Vela server deployment?
This would not affect any existing applications any way.

Author: Update these two lines is easiest

In this case, the deployment name of OAM runtime is `oam-kubernetes-runtime-oam`, let's edit it to update the image:

```shell script
$ kubectl -n oam-system edit deployment oam-kubernetes-runtime-oam
```

There are two changes:

- update the image from `crossplane/oam-kubernetes-runtime:latest` to `oamdev/vela-core:latest`
- add an args `- "--disable-caps=all"`, which will disable all additional workloads and traits built in vela-core described in the following table.

| Type | Current KubeVela Additional CRD   |
| ---- |  ----  |
| Trait | `autoscalers.standard.oam.dev` |
| Trait | `metricstraits.standard.oam.dev` |
| Workload | `podspecworkloads.standard.oam.dev` |
| Trait | `route.standard.oam.dev` |

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: oam-kubernetes-runtime-oam
...
      containers:
        - name: oam
          args:
            - "--metrics-addr=:8080"
            - "--enable-leader-election"
+           - "--disable-caps=all"
-         image: crossplane/oam-kubernetes-runtime:latest
+         image: oamdev/vela-core:latest
          imagePullPolicy: "Always"
...
```

> ### Option 2: You want use fully functioned KubeVela

Reviewer 2: ```suggestion
### Option 2: I want full functioning KubeVela, including its built-in workload types and traits.
```

> 1. Install Additional CRDs

Reviewer 3: This section doesn't talk about migration? What about existing deployment?

Author: In step 6.

$ kubectl apply -f charts/vela-core/templates/cert-manager.yaml
```

5. Install Vela Dependency ConfigMap

```shell script
$ kubectl apply -f charts/vela-core/templates/velaConfig.yaml
```

6. Delete your old oam-runtime deployment

Find the running deployment.

```shell script
$ kubectl -n oam-system get deployment -l app.kubernetes.io/name=oam-kubernetes-runtime
NAME                         READY   UP-TO-DATE   AVAILABLE   AGE
oam-kubernetes-runtime-oam   1/1     1            1           62s
```

Delete the deployment found.

```shell script
$ kubectl -n oam-system delete deployment oam-kubernetes-runtime-oam
```

7. Install Certificate and Webhook for the new controller

```shell script
$ helm template --release-name kubevela -n vela-system -s templates/webhook.yaml charts/vela-core/ | kubectl apply -f -
```

8. Install the new controller

```shell script
$ helm template --release-name kubevela -n vela-system -s templates/kubevela-controller.yaml charts/vela-core/ | kubectl apply -f -
```

> TIPS: If you want to disable webhook, change 'useWebhook' to be 'false' in  `charts/vela-core/values.yaml`

Then you have successfully migrate from oam-kubernetes-runtime to KubeVela.

> ## For Users Who are Using OAM runtime as Code Library

Reviewer 2: ```suggestion
## For Users Who are Importing OAM runtime as Code Library
```

> If you are using `oam-kubernetes-runtime` as code library, you can update your import headers.

Reviewer 2: ```suggestion
If you are importing `oam-kubernetes-runtime` as library, you can update your import headers.
```
